library/ssl_cli.c

```diff
@@ -790,125 +869,302 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
+ * used by ssl_client_hello_write(). */
+static ssl_client_hello_ext_write_func_t ext_write_funcs[] =
+{
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    ssl_write_hostname_ext,
+#endif
+#if defined(MBEDTLS_SSL_RENEGOTIATION)
+    /* Note that TLS_EMPTY_RENEGOTIATION_INFO_SCSV is always added
+     * even if MBEDTLS_SSL_RENEGOTIATION is not defined. */
+    ssl_write_renegotiation_ext,
+#endif
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2) &&                \
+    defined(MBEDTLS_KEY_EXCHANGE__WITH_CERT__ENABLED)
+    ssl_write_signature_algorithms_ext,
+#endif
+#if defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) ||      \
+    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED)
+    ssl_write_supported_elliptic_curves_ext,
+    ssl_write_supported_point_formats_ext,
+#endif
+#if defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED)
+    ssl_write_ecjpake_kkpp_ext,
+#endif
+#if defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH)
+    ssl_write_max_fragment_length_ext,
+#endif
+#if defined(MBEDTLS_SSL_TRUNCATED_HMAC)
+    ssl_write_truncated_hmac_ext,
+#endif
+#if defined(MBEDTLS_SSL_ENCRYPT_THEN_MAC)
+    ssl_write_encrypt_then_mac_ext,
+#endif
+#if defined(MBEDTLS_SSL_EXTENDED_MASTER_SECRET)
+    ssl_write_extended_ms_ext,
+#endif
+#if defined(MBEDTLS_SSL_ALPN)
+    ssl_write_alpn_ext,
+#endif
+#if defined(MBEDTLS_SSL_SESSION_TICKETS)
+    ssl_write_session_ticket_ext,
+#endif
+    NULL
+};
+
+static int ssl_client_hello_write( mbedtls_ssl_context *ssl,
+                                   unsigned char* buf,
+                                   size_t buflen,
+                                   size_t *olen )
+{
+    size_t session_id_len;
+    size_t ciphersuite_list_len; /* Length of ciphersuite list in bytes,
+                                  * including potential additions like
+                                  * SCSV fallback. */
+    size_t total_ext_len;        /* Size of list of extensions    */
+    size_t cur_ext_len;          /* Size of the current extension */
+    uint8_t offer_compress;
+
+    unsigned char *p;   /* Used during extension writing where the
+                         * buffer pointer to the beginning of the
+                         * extension list must be kept to write
+                         * the total extension list size in the end. */
+
+    unsigned char *start = buf;
+    unsigned char *end   = buf + buflen;
+
+    const int *ciphersuites;
+    const mbedtls_ssl_ciphersuite_t *ciphersuite_info;
+    size_t i; /* Used to iterate through the ciphersuite list. */
+
+    /*
+     *     0  .   0   handshake type
+     *     1  .   3   handshake length
+     *     4  .   5   highest version supported
+     *     6  .   9   current UNIX time
+     *    10  .  37   random bytes
+     */
+
+    /* NOTE:
+     * Even for DTLS, we are writing a TLS header here.
+     * The actual DTLS handshake header is inserted in
+     * the record writing routine mbedtls_ssl_write_record().
+     */
+
+    if( buflen < 4  /* handshake header */ +
```

Comment:
You could extract these 3 values to variables and the comments wouldn't be needed. Plus, it would be clear below.

Reply:
Done.